hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/utils/QuotaHelper.java

```diff
@@ -142,10 +151,76 @@ public static void assertAssignmentQuota(final long requested, final long limit,
      */
     public static void assertAssignmentRequestSizeQuota(final long requested, final long limit) {
         if (requested > limit) {
-            final String message = String.format(
-                    "Quota exceeded: Cannot assign %s entities at once. The maximum is %s.", requested, limit);
+            final String message = String.format(MAX_ASSIGNMENT_QUOTA_EXCEEDED, requested, limit);
             LOG.warn(message);
             throw new QuotaExceededException(message);
         }
     }
+
+    /**
+     * Assert that the size of a single artifact does not exceed the limit
+     *
+     * @param filename
+     *            name of the artifact, used in logs
+     * @param softwareModuleId
+     *            id of the software module, used on logs
+     * @param artifactSize
+     *            size of the artifact
+     * @param maxArtifactSize
+     *            max allowed file size
+     */
+    public static void assertMaxArtifactSizeQuota(final String filename, final long softwareModuleId,
+            final long artifactSize, final long maxArtifactSize) {
+        if (maxArtifactSize <= 0) {
+            return;
+        }
+        if (artifactSize > maxArtifactSize) {
+            final String msg = String.format(MAX_ARTIFACT_SIZE_EXCEEDED, filename,
+                    byteValueToReadableString(artifactSize), softwareModuleId,
+                    byteValueToReadableString(maxArtifactSize));
+            LOG.warn(msg);
+            throw new QuotaExceededException(msg, QuotaExceededException.QuotaType.SIZE_QUOTA);
+        }
+    }
+
+    /**
+     * Assert that the size of an artifact does not exceed the allowed total
+     * artifact storage size
+     *
+     * @param filename
+     *            name of the artifact, used in logs
+     * @param artifactSize
+     *            size of the artifact
+     * @param currentlyUsed
+     *            currently occupied artifact storage
+     * @param maxArtifactSizeTotal
+     *            max allowed total artifact storage size
+     */
+    public static void assertMaxArtifactStorageQuota(final String filename, final long artifactSize,
+            final long currentlyUsed, final long maxArtifactSizeTotal) {
+        if (maxArtifactSizeTotal <= 0) {
+            return;
+        }
+
+        if (currentlyUsed + artifactSize > maxArtifactSizeTotal) {
+            final String msg = String.format(MAX_ARTIFACT_SIZE_TOTAL_EXCEEDED, filename,
+                    byteValueToReadableString(artifactSize), byteValueToReadableString(maxArtifactSizeTotal));
+            LOG.warn(msg);
+            throw new QuotaExceededException(msg, QuotaExceededException.QuotaType.SIZE_QUOTA);
+        }
+    }
+
+    /**
+     * Convert byte values to human readable strings with units
+     */
+    private static String byteValueToReadableString(long byteValue) {
```

Comment:
Check if part of used library

Reply:
Apache io fileutils has rounding issues (see https://issues.apache.org/jira/browse/IO-373), therefore we keep the simple local solution.